### ISSUE TITLE
无法复现论文中未迁移的RUL结果

### ISSUE THREAD

**Issue by myalos** [HAIRLAB/Health_status_prediction#2]
感谢分享数据和代码，我用这个repo里面提供的权重在测试集上面评估，发现评估出来的结果和论文中差异比较明显
下面是我的测试代码，如果有问题希望能得到指正

```python
# new_test  = ['9-6','4-5','1-2', '10-7','1-1', '6-1','6-6', '9-4','10-4','8-5', '5-3','10-6',
#            '2-5','6-2','3-1','8-8', '8-1','8-6','7-6','6-8','7-5','10-1']
from tqdm import tqdm
# 新的评估方法
model = CRNN(100,4,64,64)
model = model.to('cuda')
model.load_state_dict(torch.load('model/wx_inner/fromgithub.pt')) # 这个是从github上面下的权重
trainer = Trainer(lr = 8e-4, n_epochs = None,device = device, patience = 1200,
                  lamda = lamda, alpha = None, model_name=None)
ape = [] # 论文中所有cycle的ape

for name in tqdm(new_test):
    tmp_fea, tmp_lbl = all_loader[name]['fea'],all_loader[name]['lbl']
    datalength = len(tmp_fea)
    filename = Path('../../Data/hust/') / f'{name}.pkl'
    cyclelife = pickle.load(open(filename, 'rb'))[name]['rul'][1] + 1
    test_fea_ = tmp_fea[:].copy()
    test_lbl_ = tmp_lbl[:].copy()
    test_fea_ = test_fea_.transpose(0,3,2,1)
    testset = TensorDataset(torch.Tensor(test_fea_), torch.Tensor(test_lbl_))
    test_loader = DataLoader(testset, batch_size=len(testset), shuffle=False, drop_last = False)
    y_true, y_pred, _, _, soh_pred = trainer.test(test_loader, model)
    y_true = y_true.cpu().detach().numpy() * rul_factor
    y_pred = y_pred.cpu().detach().numpy() * rul_factor
    ape.append(np.mean(np.abs(y_true - y_pred)) / cyclelife * 100)
print(ape)
# 输出
# [39.36477284696823, 43.46154441296215, 31.333708611830442, 31.474831072204317, 30.941896153446848, 29.73903111049107, 30.637107160034425, 32.87269808502815, 33.022902869757175, 32.7041026483738, 31.14490608300418, 34.418119762617565, 48.637393437105715, 31.454437399734807, 33.52355799643018, 42.75005885532924, 37.21408130013369, 32.47438948333818, 38.84136790960607, 32.11363127454355, 33.56998337865638, 34.799955931398635]
print(np.mean(ape))
# 34.84 % 和论文中的 8.84% 差的有点多
```

**Comment by mxt0607** [comment]
Thanks. Our results can be reproduced smoothly. According to your code, the dimensions of arrays are different between "y_true" and "y_pred," resulting in the computation of an average over a matrix.

**Comment by mxt0607** [comment]
Thanks. Our results can be reproduced smoothly. According to your code, the dimensions of arrays are different between "y_true" and "y_pred", resulting in the computation of an average over a matrix.

**Comment by myalos** [comment]
You are right, thanks for reply and correction! After the correction, the result become 8.39%, and it seems to be better than the result in the paper. The following is my evaluation code:
```python
# change data path, from data/our_data to our_data   
# The path in get_xy function also need be changed accordingly
import os
import time
import pickle
import math
import random
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.pyplot import MultipleLocator
import seaborn as sns
from scipy import interpolate
from datetime import datetime
import pandas as pd
from tool import EarlyStopping
from sklearn.metrics import roc_auc_score,mean_squared_error,mean_absolute_error, r2_score
from common import *
from net import CRNN
import torch
import torch.nn.functional as F
import torch.optim as optim

from torch import nn
from torch.autograd import Variable
from torch.utils.data import DataLoader, Dataset, Sampler, TensorDataset
from torch.utils.data.sampler import RandomSampler

new_test  = ['9-6','4-5','1-2', '10-7','1-1', '6-1','6-6', '9-4','10-4','8-5', '5-3','10-6',
            '2-5','6-2','3-1','8-8', '8-1','8-6','7-6','6-8','7-5','10-1']
n_cyc = 30
in_stride = 3
fea_num = 100

v_low = 3.36
v_upp = 3.60
q_low = 610
q_upp = 1190
rul_factor = 3000
cap_factor = 1190
pkl_list = os.listdir('our_data/')
pkl_list = sorted(pkl_list, key=lambda x:int(x.split('-')[0])*10 + int(x[-5]))

train_name = []
for name in pkl_list:
    train_name.append(name[:-4])
all_loader = dict()
print('----init_train----')
for name in train_name:
    tmp_fea, tmp_lbl = get_xy(name, n_cyc, in_stride, fea_num, v_low, v_upp, q_low, q_upp, rul_factor, cap_factor)
    all_loader.update({name:{'fea':tmp_fea,'lbl':tmp_lbl}})

n_cyc = 30
in_stride = 3
fea_num = 100

v_low = 3.36
v_upp = 3.60
q_low = 610
q_upp = 1190
rul_factor = 3000
cap_factor = 1190
pkl_list = os.listdir('our_data/')
pkl_list = sorted(pkl_list, key=lambda x:int(x.split('-')[0])*10 + int(x[-5]))
all_loader = pickle.load(open('allloader.pkl', 'rb'))

from tqdm import tqdm
lamda = 0.0
train_weight9 = [0., 0.1, 0., 0., 0.1, 0., 0., 0., 0.,]
valid_weight9 = [0. if (i!=0) else 0.1 for i in train_weight9]
train_alpha = torch.Tensor(train_weight9 + [0.] )
valid_alpha = torch.Tensor(valid_weight9 + [0.])
device = 'cuda'

# 新的评估方法
model = CRNN(100,4,64,64)
model = model.to('cuda')
model.load_state_dict(torch.load('model/wx_inner/wx_inner_pretrain_end.pt')) 
trainer = Trainer(lr = 8e-4, n_epochs = None,device = 'cuda', patience = 1200,
                  lamda = lamda, alpha = None, model_name=None)
ape = [] # 论文中所有cycle的ape

for name in tqdm(new_test):
    tmp_fea, tmp_lbl = all_loader[name]['fea'],all_loader[name]['lbl']
    datalength = len(tmp_fea)
    filename = Path('our_data/') / f'{name}.pkl'
    cyclelife = pickle.load(open(filename, 'rb'))[name]['rul'][1] + 1
    test_fea_ = tmp_fea[:].copy()
    test_lbl_ = tmp_lbl[:].copy()
    test_fea_ = test_fea_.transpose(0,3,2,1)
    testset = TensorDataset(torch.Tensor(test_fea_), torch.Tensor(test_lbl_))
    test_loader = DataLoader(testset, batch_size=len(testset), shuffle=False, drop_last = False)
    y_true, y_pred, _, _, soh_pred = trainer.test(test_loader, model)
    y_true = y_true.cpu().detach().numpy() * rul_factor
    y_pred = y_pred.cpu().detach().numpy() * rul_factor
    ape.append(np.mean(np.abs(y_true - y_pred.squeeze())) / cyclelife * 100)
print(ape)
# output 15.104215690888553, 19.140458931232995, 6.7525825913426765, 2.5150305487115285, 3.9791474706706813,7.119564743515867,14.025065598578626, 2.398954036264767, 3.450334730000949,1.6840634300056612, 5.6313219034982, 4.625765649456782, 29.367622094955387, 2.9755384026552885,
    #6.895486124401673,18.028299240838912,10.210277070700439,2.686927002159499,12.168382456604864,6.460481018788568,
#    2.829773542977599,6.547118997265855
print(np.mean(ape)) # 8.39
```

**Comment by zhudengliang** [comment]
I have a question, why do you use the test set data and labels to fine-tune training during the fine-tuning process, and then use the test set to test on the fine-tuned model? Does that make any sense? Or am I mistaken? Here's the code from '1-wx_inner'

        trainer = FineTrainer(lr = 1e-4, n_epochs = num_epochs,device = device, patience = 1000,
                      lamda = lamda, train_alpha = train_alpha, valid_alpha = valid_alpha, model_name=finetune_model_path)
        model ,train_loss, valid_loss, total_loss, added_loss = trainer.train(test_loader, test_loader, model, model_load)

        y_true, y_pred, mse_loss, soh_true, soh_pred = trainer.test(test_loader, model)

**Comment by mxt0607** [comment]
Thank you for your question. The testing process involves fine-tuning the pre-trained model using charging data and SOHs from the (N-9th to N-1th cycles) for each testing charge-discharge cycle (Nth cycle) of the battery. The current SOH and RUL for the Nth cycle are then estimated and predicted based on the labeled SOHs and charging data from the previous cycles.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "HAIRLAB/Health_status_prediction"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年12月25日(星期三) 上午10:45
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [HAIRLAB/Health_status_prediction] 无法复现论文中未迁移的RUL结果 (Issue #2)





 
I have a question, why do you use the test set data and labels to fine-tune training during the fine-tuning process, and then use the test set to test on the fine-tuned model? Does that make any sense? Or am I mistaken? Here's the code from '1-wx_inner'
     trainer = FineTrainer(lr = 1e-4, n_epochs = num_epochs,device = device, patience = 1000,                   lamda = lamda, train_alpha = train_alpha, valid_alpha = valid_alpha, model_name=finetune_model_path)     model ,train_loss, valid_loss, total_loss, added_loss = trainer.train(test_loader, test_loader, model, model_load)     y_true, y_pred, mse_loss, soh_true, soh_pred = trainer.test(test_loader, model)
